### PR TITLE
envision: autopatchelf after build

### PR DESCRIPTION
**Pull request by Pandapip1**
## Description of changes

Adds an OpenXR document to establish an expectation (and a little bit of responsibility), and updates envision to align with that new document. This is necessary because otherwise all OpenXR applications would need to run in a fhsenv with a list of commonly-used provider dependencies.

Also makes envision align with said document.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Pandapip1** [comment]
I'm aware `autopatchelfHook` exists, but it doesn't actually expose the `autopatchelf` binary.

**Comment by Artturin** [comment]
> I'm aware `autopatchelfHook` exists, but it doesn't actually expose the `autopatchelf` binary.

Maybe expose it in the `passthru` of `autoPatchelfHook` if its maintainers are ok with that.

The old `autopatchelf` needs to be clearly labelled as `old-autopatchelf` to make sure no one uses it.

**Comment by Pandapip1** [comment]
> The old autopatchelf needs to be clearly labelled as old-autopatchelf to make sure no one uses it. (preferably it's not added to nixpkgs)

Out of curiosity, why shouldn't it be used?

**Comment by Artturin** [comment]
> > The old autopatchelf needs to be clearly labelled as old-autopatchelf to make sure no one uses it. (preferably it's not added to nixpkgs)
> 
> Out of curiosity, why shouldn't it be used?

It's likely lacking all the fixes the in-tree auto-patchelf.sh had before being rewritten in python and the fixes the python version has

**Comment by Scrumplex** [comment]
> Maybe expose it in the `passthru` of `autoPatchelfHook` if its maintainers are ok with that.

I have proposed #340162 for this ^^

**Comment by ElvishJerricco** [comment]
@Pandapip1

> I'm aware autopatchelfHook exists, but it doesn't actually expose the autopatchelf binary.

Because it doesn't use it. That binary is obsolete and what we use now is in-tree.

**Comment by Pandapip1** [comment]
Yeah, @Scrumplex already told me that. I'm reviewing their PR to expose it.

**Comment by Pandapip1** [comment]
Tracking @Scrumplex's exposing of auto-patchelf: https://nixpk.gs/pr-tracker.html?pr=340162

**Comment by Pandapip1** [comment]
Bump. Would like to get this merged in time for 24.11

File: pkgs/by-name/au/auto-patchelf/source/auto-patchelf.py [hunk 1]
```diff
@@ -394,14 +413,21 @@ def main() -> None:
         action="store_true",
         help="Attempt to search for and relink libc dependencies.",
     )
+    parser.add_argument(
+        "--no-add-existing",
```
Comment (layus) [inline pkgs/by-name/au/auto-patchelf/source/auto-patchelf.py:417]: I think the name does not convey the right meaning. I would prefer something like `--no-default-paths` or `--no-self-references`.

The description below can be confusing to the readers too. The thing that is avoided here is adding the paths to patch to the search paths. Ir makes sense as you are supposed to prefer binaries that you are providing over system ones.

That is why I am a bit uncertain about this whole change. Until now this problem has always been solved by removing the libraries that autopatchelf should not consider, or by passing files in paths_to_patch.

The patch is well written, and there is no problem with it _per se_, but I would like to understand why and if it is really needed.

File: pkgs/by-name/en/envision/package.nix [hunk 2]
```diff
@@ -1,45 +1,36 @@
 {
+  lib,
```
Comment (layus) [inline pkgs/by-name/en/envision/package.nix:2]: nit: this file does not pass the formatting check on CI

**Review by layus — CHANGES_REQUESTED** [review]
Thanks for pushing this forward, it is a very interesting idea that opens the way to more self-updating packages. Could you answer the questions in my above (below?) comment ? I promise I will be more responsive to interact and ultimately merge.

File: pkgs/by-name/au/auto-patchelf/source/auto-patchelf.py (hunk 1 above)
Comment (Pandapip1) [reply pkgs/by-name/au/auto-patchelf/source/auto-patchelf.py:417]: > The patch is well written, and there is no problem with it *per se*, but I would like to understand why and if it is really needed.

The reason this is needed is that in an fhsenv (such as the envision build environment), auto-patchelf will find the libraries in the fhsenv due to the binaries' rpaths, and won't modify the binaries that need the rpaths to point into the nix store.

File: pkgs/by-name/en/envision/package.nix (hunk 2 above)
Comment (layus) [reply pkgs/by-name/en/envision/package.nix:2]: Would you mind fixing the formatting. I can also do it myself if you prefer.

**Comment by Pandapip1** [comment]
Sorry all; realized my mistake right as I hit enter on the force push.

**Pull request closed without merging**